### PR TITLE
Reduce memory used to store inactive policies/profiles

### PR DESCRIPTION
**Pull request by fasaxc**
## Description

The `Rule` struct that is used in policies and profiles has grown very large over time.  With allocation overheads, policies end up using approximately 1KB per rule!

We were previously storing pointers to the policies in two places.  We need to change both in order to realise any savings:

- The active rules calculator needs the full policy struct to be available when it determines that a policy is active.
- The policy resolver/policy sorter only need to know the name and order of the policy.

We tackle them in two different ways:

- In the active rules calculator, store inactive policies in a compressed format; first encoding them as JSON, then compressing with "snappy" (a fast compression algorithm).  The data is already marked up for JSON serialisation (as used by Typha to serialise the stream to Felix) so this should be very safe.  Unpack the policies on demand. Notes:

  - Unpacks should happen rarely, only when a policy goes from inactive to active.
  - We could optimise the storage of active policies in future by changing from a sparse struct to something like a slice of match criteria.  But using the encoding we already had seemed like an easy win.

- In the policy resolver, avoid storing the whole policy and extract the needed metadata in a compact struct.  Since we're transforming the data anyway, map policy order "default" to float `+Inf`.  This sorts exactly as desired and avoids awkward special case logic.

I measured the saving using 1,000 namespaces, each with one k8s NetworkPolicy that allowed DNS traffic (UDP and TCP to a handful of CIDRs), traffic within the namespace and traffic from an external namespace.  (This was a policy supplied by a user who was running into occupancy problems with 10s of thousands of namespaces.)

* Before the changes, each namespace+policy used ~25KiB (heap inuse space).
* After the changes, that's down to 7.2KiB.

## Todos

- [x] Tests
- [x] Documentation
- [x] Release note

## Release Note

```release-note
Felix now compresses network policies and namespaces in memory, reducing RAM usage in clusters with many policies/namespaces that are not active on a given node.
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: libcalico-go/lib/packedmap/packedmap.go
Comment (tomastigera) [inline libcalico-go/lib/packedmap/packedmap.go:129]: packed would be a subslice of arr if arr was large enough. Does it mean that it always takes 16k at least?

**Review by tomastigera — APPROVED** [review]
LGTM with one question, looks like s straight forward refactoring

File: libcalico-go/lib/packedmap/packedmap.go
Comment (fasaxc) [reply libcalico-go/lib/packedmap/packedmap.go]: Tweaked it to re-use the same cached buffer over and over.  That will force the compiler to allocate the string, as desired.  (Right now, the go compiler does allocate but I couldn't find anything in the language spec to say it always will.)

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and leave the branch after I've merged it.

**Pull request merged**
